free/src/test/scala-2.13+/cats/free/FreeStructuralSuite.scala

```diff
@@ -0,0 +1,133 @@
+package cats.free
+
+import cats.{Applicative, Eq, Eval, Functor, Show, Traverse}
+import cats.kernel.laws.discipline.{EqTests, /*HashTests,*/ PartialOrderTests}
+import cats.syntax.all._
+import cats.tests.CatsSuite
+
+import org.scalacheck.{Arbitrary, Cogen, Gen}
+
+// this functionality doesn't exist on Scala 2.12
+class FreeStructuralSuite extends CatsSuite {
+  import FreeSuite.freeArbitrary
+  import FreeStructuralSuite._
+
+  implicit def freeCogen[S[_]: Functor, A](implicit S: => Cogen[S[Free[S, A]]], A: Cogen[A]): Cogen[Free[S, A]] =
+    Cogen { (seed, f) =>
+      f.resume match {
+        case Left(sf) =>
+          S.perturb(seed, sf)
+
+        case Right(a) =>
+          A.perturb(seed, a)
+      }
+    }
+
+  Show[Free[Option, Int]]
+
+  // TODO HashLaws#sameAsUniversalHash is really dodgy
+  // checkAll("Free[Option, Int]", HashTests[Free[Option, Int]].hash)
+  checkAll("Free[Option, Int]", PartialOrderTests[Free[Option, Int]].partialOrder)
```

Comment:
is it the case that `f1: Free[Option, Int] == f2: Free[Option, Int] <=> f1.runTailRec == f2.runTailRec`?

That seems like a nice property to have, but I don't see that if the current code can get it.

If it can't get it, what does this notion of equality mean?

Reply:
I guess if I read the docs I might learn something...

So, I guess the property I mention is not true, but one way: if we are "free-equal" then we are option-equal.

That might be a nice test to add to strengthen the tests here (e.g. I think returning that all items are equivalent always passes the laws, but isn't what we really want, adding a law I suggest here rules out many bad implementations since happening to get equal `Option[Int]` is unlikely.

Reply:
> So, I guess the property I mention is not true, but one way: if we are "free-equal" then we are option-equal.

I *believe* this is the case. Caffeine is still working its way into my brain, but at first glance, I think this property holds*at least* because `Either` is monotone with respect to `Eq` (meaning that `a1 === a2` implies `Right(a1) === Right(a2)` and `Left(a1) === Left(a2)` *and* the converse), which means that this would reduce to an inductive argument over that monotonicity. But, as I'll argue in a minute, this restriction on the pattern functor is probably not needed.

The stronger bidirectional implication definitely doesn't hold, since `Free.pure(42).runTailRec === Free.suspend(Some(42)).runTailRec`, but they would *not* be considered equal by the structural definition of `Eq`. You can build versions of this which don't rely on coyoneda as well. Interpreting a `Free` into a monad is a stronger process than simply inspecting its defining structure, which is what this `Eq` does. In a sense, this is the reification of the intuition that, for almost any value, there are many different programs which can produce that outcome. The structural `Eq` inspects the *program*, while `runTailRec` inspects the *outcome*.

What you're saying is that, for any *programs* that are structurally equal, the *outcomes* they produce when run must also be equal. I believe this holds for all suspension functors which are deterministic monads (e.g. `race` makes `IO` a nondeterministic monad, which is why it is generally ignored as a combinator).